crates/flowctl/src/raw/suggest_schema.rs

```diff
@@ -0,0 +1,199 @@
+            "last_pub_user_id",
+            "spec_type",
+            "spec",
+        ],
+    )
+    .await?;
+
+    let (_, collection_def) = collect_specs(live_specs)?.collections.pop_first().ok_or(anyhow!("could not find collection"))?;
+
+    // Reader for the collection itself
+    let selector = CollectionJournalSelector {
+        collection: collection.clone(),
+        ..Default::default()
+    };
+    let reader = journal_reader(ctx, &ReadArgs {
+        selector,
+        ..Default::default()
+    }).await?;
+
+    // Reader for the ops log of the task
+    let ops_collection = "ops.us-central1.v1/logs".to_string();
+    let selector = CollectionJournalSelector {
+        collection: ops_collection.clone(),
+        include_partitions: vec![Partition {
+            name: "name".to_string(),
+            value: task.clone(),
+        }, Partition {
+            name: "kind".to_string(),
+            value: "capture".to_string(),
+        }],
+        ..Default::default()
+    };
+
+    // Read log lines from the logs collection and filter "failed validation" documents
+    let log_reader = journal_reader(ctx, &ReadArgs {
+        selector,
+        ..Default::default()
+    }).await?;
+    let log_stream = AsyncByteLines::new(BufReader::new(log_reader.compat())).into_stream();
+    let log_invalid_documents = log_stream.try_filter_map(|log| async move {
+        let parsed: Log = serde_json::from_slice(&log)?;
+        if parsed.message != "document failed validation against its collection JSON Schema" {
+            return Ok(None);
+        }
+
+        let error_string = &parsed.fields_json_map.get("error").ok_or(to_io_error("could not get 'error' field of ops log"))?;
+        let mut err: Vec<serde_json::Value> = serde_json::from_str(error_string)?;
+        let err_object = err.pop().ok_or(to_io_error("could not get second element of 'error' field in ops log"))?;
+        let failed_validation: FailedValidation = serde_json::from_value(err_object)?;
+
+        return Ok(Some(failed_validation.document));
+    });
+
+    let codec = JsonCodec::new(); // do we want to limit length here? LinesCodec::new_with_max_length(...) does this
+
+    // Chain together the collection document reader and the log_invalid_documents stream so we can
+    // run schema-inference on both
+    let mut docs_stream = Box::pin(FramedRead::new(FuturesAsyncReadCompatExt::compat(reader), codec).map_err(to_io_error).chain(log_invalid_documents));
+
+    // The original collection schema to be used as the starting point of schema-inference
+    let schema_model = collection_def.schema.unwrap();
+    // The inferred shape, we start by using the existing schema of the collection
+    let mut inferred_shape = raw_schema_to_shape(&schema_model)?;
+
+    // Create a JSONSchema object from the original schema so we can use it to run a diff later
+    let original_jsonschema = SchemaBuilder::new(inferred_shape.clone()).root_schema();
```

Comment:
When you say lossy, you mean in terms of the original JSON structure of the file, right? Not in terms of the meaning of the jsonschema.

The reason for this is that we want to have two JSONSchema structures that are similar in layout, which would allow us to run a more intuitive diff, and as long as the meaning of the JSONSchema stays the same, this diff would be informative for the user to be able to write a JSON Merge Patch.